core/server/worker/src/main/java/alluxio/worker/block/CacheRequestManager.java

```diff
@@ -0,0 +1,345 @@
+import alluxio.worker.block.io.BlockWriter;
+
+import com.codahale.metrics.Counter;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Handles client requests to synchronously/asynchronously cache blocks. Responsible for
+ * managing the local worker resources and intelligent pruning of duplicate or meaningless requests.
+ */
+@ThreadSafe
+public class CacheRequestManager {
+  private static final Logger LOG = LoggerFactory.getLogger(CacheRequestManager.class);
+  private static final Logger SAMPLING_LOG = new SamplingLogger(LOG, 10L * Constants.MINUTE_MS);
+
+  /** Executor service for execute the async cache tasks. */
+  private final ExecutorService mCacheExecutor;
+  /** The block worker. */
+  private final BlockWorker mBlockWorker;
+  private final ConcurrentHashMap<Long, CacheRequest> mActiveCacheRequests;
+  private final String mLocalWorkerHostname;
+  private final FileSystemContext mFsContext;
+  /** Keeps track of the number of rejected cache requests. */
+  private final AtomicLong mNumRejected = new AtomicLong(0);
+
+  /**
+   * @param service thread pool to run the background caching work
+   * @param blockWorker handler to the block worker
+   * @param fsContext context
+   */
+  public CacheRequestManager(ExecutorService service, BlockWorker blockWorker,
+      FileSystemContext fsContext) {
+    mCacheExecutor = service;
+    mBlockWorker = blockWorker;
+    mFsContext = fsContext;
+    mActiveCacheRequests = new ConcurrentHashMap<>();
+    mLocalWorkerHostname = NetworkAddressUtils.getLocalHostName(
+        (int) ServerConfiguration.getMs(PropertyKey.NETWORK_HOST_RESOLUTION_TIMEOUT_MS));
+  }
+
+  /**
+   * Handles a request to cache a block. If it's async cache, it wouldn't throw exception.
+   *
+   * @param request the cache request fields will be available
+   */
+  public void submitRequest(CacheRequest request)
+      throws AlluxioException, IOException {
+    CACHE_REQUESTS.inc();
+    long blockId = request.getBlockId();
+    boolean async = request.getAsync();
+    if (mActiveCacheRequests.putIfAbsent(blockId, request) != null) {
+      // This block is already planned and just just return.
+      if (async) {
+        LOG.debug("request already planned: {}", request);
+      } else {
+        try {
+          CommonUtils.waitFor("block to be loaded",
+              () -> mActiveCacheRequests.containsKey(blockId));
```

Comment:
shall we put a timeout option here or it will block forever in `waitFor`?

Reply:
added wait option for 30s. catch a bug in the meantime, update to !mActiveCacheRequests.containsKey(blockId)